knodle/trainer/snorkel/snorkel.py

```diff
@@ -22,19 +25,29 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
     def _snorkel_denoising(self, model_input_x, rule_matches_z):
+        """
+        Premise:
+            Snorkel can not make use of rule-unlabeled examples (no rule matches).
+            The generative LabelModel assigns such examples a uniform distribution over all available labels,
+            which contradicts the desired behaviour. Such examples should be either filtered or assigned an
+            "other class id".
+        Filtering / other class strategy:
+            filter_non_labelled = True:
+                Drop the unlabeled examples completely prior to LabelModel.
+            filter_non_labelled = False:
+                However, we might want to keep negative examples in the training data, but we should not pass them
+                through the LabelModel. Therefore, the rule-unlabeled part of the data skips the LabelModel step
+                and is added directly to the output data with manually assigned "other class id".
+        """
 
         # initialise optimizer
         self.trainer_config.optimizer = self.initialise_optimizer()
 
-        non_zero_indices = np.where(rule_matches_z.sum(axis=1) != 0)[0]
-        rule_matches_z = rule_matches_z[non_zero_indices]
-        tensors = list(model_input_x.tensors)
-        for i in range(len(tensors)):
-            tensors[i] = tensors[i][non_zero_indices]
-
-        model_input_x = TensorDataset(*tensors)
-
         # create Snorkel matrix and train LabelModel
+        non_empty_mask, rule_matches_z, model_input_x = prepare_empty_rule_matches(
```

Comment:
move filtering of `model_input_x` matrix from this function to line 65 where the whole output data is adjusted for preserving non-labelled samples?